nshlib/nsh_parse.c

```diff
@@ -2647,6 +2673,81 @@ static int nsh_parse_command(FAR struct nsh_vtbl_s *vtbl, FAR char *cmdline)
           param.oflags_in       = O_RDONLY;
           param.file_in         = nsh_getfullpath(vtbl, arg);
         }
+#ifdef CONFIG_NSH_PIPELINE
+      else if (!strncmp(argv[argc], g_pipeline1, pipeline1_len))
+        {
+          FAR char *arg;
+          FAR char *sh_argv[MAX_ARGV_ENTRIES];
+
+          if (argv[argc][pipeline1_len])
+            {
+              arg = &argv[argc][pipeline1_len];
+            }
+          else
+            {
+              arg = nsh_argument(vtbl, &saveptr, &memlist, NULL, &isenvvar);
+            }
+
+          if (!arg)
+            {
+              nsh_error(vtbl, g_fmtarginvalid, cmd);
+              ret = ERROR;
+              goto dynlist_free;
+            }
+
+          argv[argc] = NULL;
+          sh_argv[0] = "sh";
+          sh_argv[1] = "-c";
+          for (ret = 0; ret <= argc; ret++)
+            {
+              sh_argv[ret + 2] = argv[ret];
+            }
+
+          ret = pipe2(pipefd, 0);
+          if (ret < 0)
+            {
+              ret = -EPIPE;
+              goto dynlist_free;
+            }
+
+          redirect_out_save = vtbl->np.np_redir_out;
+          vtbl->np.np_redir_out = true;
+          param.fd_out = pipefd[1];
+
+          bg_save = vtbl->np.np_bg;
+          vtbl->np.np_bg = true;
+
+          ret = nsh_execute(vtbl, argc + 2, sh_argv, &param);
+
+          vtbl->np.np_bg = bg_save;
+
+          if (param.fd_in != -1)
+            {
+              close(param.fd_in);
+              param.fd_in = -1;
+              vtbl->np.np_redir_in = redirect_in_save;
+            }
+
+          if (param.fd_out != -1)
+            {
+              close(param.fd_out);
+              param.fd_out = -1;
+              vtbl->np.np_redir_out = redirect_out_save;
+            }
+
+          if (ret == -1)
```

Comment:
move after line 2748 to ensure param.fd_in get closed in the fail path

Reply:
Done.